File: code/itsnowkim/week6/prog_118668.py
# 모든 문제들을 풀 수 있는 alp, cop 를 얻는 최단시간은?
def solution(alp, cop, problems):
    # 도달해야 하는 alp, cop
    req_max_alp = max([x[0] for x in problems])
    req_max_cop = max([x[1] for x in problems])
    
    INF = 10e9

    # 이미 목표치보다 커지는 경우 예외 처리
    alp = min(alp, req_max_alp)
    cop = min(cop, req_max_cop)
    

    # 해당 alg, cop 를 얻기까지 소요되는 시간을 저장
    dp = [[INF for _ in range(req_max_cop+1)] for _ in range(req_max_alp+1)]
    
    # init
    dp[alp][cop] = 0
    
    # dp table update
    for i in range(alp, req_max_alp+1):
        for j in range(cop, req_max_cop+1):
            # 알고력 높이기
            # 존재하지 않는 index일 경우 수행 x
            if i+1 <= req_max_alp:
                dp[i+1][j] = min(dp[i+1][j], dp[i][j]+1)

            # 코딩력 높이기
            if j+1 <= req_max_cop:
                dp[i][j+1] = min(dp[i][j+1], dp[i][j]+1)

            # 문제 풀기
            for p in problems:
                alp_req, cop_req, alp_rwd, cop_rwd, cost = p
                if i>= alp_req and j>=cop_req: # 풀 수 있음
                    # 필요한 범위보다 커지는 경우에 처리
                    ceil_alp,ceil_cop = min(req_max_alp,i + alp_rwd), min(req_max_cop,j + cop_rwd)
                    dp[ceil_alp][ceil_cop] = min(dp[ceil_alp][ceil_cop], dp[i][j]+cost)

    # for d in dp:
        # print(d)
    return dp[-1][-1]

#15
print(solution(10, 10, [[10,15,2,1,2],[20,20,3,3,4]]))

# 13
print(solution(0,0, [[0,0,2,1,2],[4,5,3,1,2],[4,11,4,0,2],[10,4,0,4,2]]))

File: code/SungHo-Ck/week6/prog_118668.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>
#define INF 987654321

using namespace std;

int solution(int alp, int cop, vector<vector<int>> problems) {
    int answer = 0;
    int alp_need = alp;
    int cop_need = cop;
    for (int i = 0; i < problems.size(); i++) {
        alp_need = max(alp_need, problems[i][0]);
        cop_need = max(cop_need, problems[i][1]);
    }
    vector<vector<int>>dp(alp_need + 1, vector<int>(cop_need + 1, INF));
    
    alp = min(alp, alp_need);
    cop = min(cop, cop_need);
    dp[alp][cop] = 0;

    for (int i = alp; i <= alp_need; i++) {
        for (int j = cop; j <= cop_need; j++) {
            if (i < alp_need) {
                dp[i + 1][j] = min(dp[i + 1][j], dp[i][j] + 1);
            }
            if (j < cop_need) {
                dp[i][j + 1] = min(dp[i][j + 1], dp[i][j] + 1);
            }
            for (int k = 0; k < problems.size(); k++) {
                if (i >= problems[k][0] && j >= problems[k][1]) {
                    int tmp_a = min(i + problems[k][2], alp_need);
                    int tmp_c = min(j + problems[k][3], cop_need);
                    dp[tmp_a][tmp_c] = min(dp[tmp_a][tmp_c], dp[i][j] + problems[k][4]);
                }
            }
        }
    }

    return dp[alp_need][cop_need];
}

int main(void) {
    vector<vector<int>> problems = { {10,15,2,1,2} ,{20,20,3,3,4} };
    int ans = solution(10,10, problems);
    cout << ans << endl;

    problems = { {0,0,2,1,2} ,{4,5,3,1,2},{4,11,4,0,2},{10,4,0,4,2} };
    ans = solution(0,0, problems);
    cout << ans << endl;
    return 0;
}